File: CN/Experiment 3/HammingCode.c
#include <stdio.h>

char data[4], check[3], final[7], received[7];
int parity_pos[] = {3, 1, 0};
int data_pos[] = {6, 5, 4, 2};

void calculate_parity_bits() {
    for (int i = 0; i < 3; i++) {
        check[i] = 0;
        for (int j = 0; j < 4; j++) {
            if (data[j] == '1' && (j & (1 << parity_pos[i]))) {
                check[i] ^= 1;
            }
        }
        final[parity_pos[i]] = check[i] + '0';
    }
}

void print_hamming_code() {
    printf("The Hamming code is: ");
    for (int i = 0; i < 7; i++) {
        printf("%c", final[i]);
    }
    printf("\n");
}

void check_received_hamming_code(char received[]) {
    int error_position = -1;
    int calculated_check[3] = {0, 0, 0};
    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 7; j++) {
            if (received[j] == '1' && (j & (1 << parity_pos[i]))) {
                calculated_check[i] ^= 1;
            }
        }
    }
    for (int i = 0; i < 3; i++) {
        if (calculated_check[i] != (received[parity_pos[i]] - '0')) {
            error_position = (1 << parity_pos[i]) - 1;
            break;
        }
    }
    if (error_position != -1) {
        printf("Error detected at position %d\n", error_position);
        received[error_position] ^= 1; // Invert the bit at the error position
        printf("Corrected Hamming code: ");
        for (int i = 0; i < 7; i++) {
            printf("%c", received[i]);
        }
        printf("\n");
    } else {
        printf("No errors detected\n");
    }
}

int main() {
    printf("Enter the data: ");
    scanf("%s", data);
    calculate_parity_bits();
    for (int i = 0; i < 4; i++) {
        final[data_pos[i]] = data[i];
    }
    print_hamming_code();
    printf("Enter the received bit string: ");
    scanf("%s", received);
    check_received_hamming_code(received);

    return 0;
}

File: CN/Experiment 3/HammingCode.py
def calculateParityBits(n):
	i = 0
	while 2**i < n + i + 1:
		i += 1
	return i

def returnParityBits(n):
	i = 0
	while 2**i < n + 1:
		i += 1
	return i

def generateParityBits(data, check_bit, parity):
	count = 0
	for i in range(len(data)):
		if (len(data) - i) & check_bit:
			count += int(data[i])
	if parity == 0:
		if count % 2 == 0:
			return 0
		else:
			return 1
	else:
		if count % 2 == 0:
			return 1
		else:
			return 0

# Generator
data = input("Enter data: ")
noOfParity = calculateParityBits(len(data))
print(f"Number of parity bits required: {noOfParity}")
parity = int(input("Enter parity (0 for even, 1 for odd): "))
parity_pos = [2**i for i in range(noOfParity)]
final = ['0'] * (len(data) + noOfParity)
j = 1
for i in range(1, len(final) + 1):
	if i in parity_pos:  # Check if i is a power of 2
		continue
	final[-i] = data[-j]
	j += 1

for i in range(noOfParity):
	final[-parity_pos[i]] = str(generateParityBits(final, parity_pos[i], parity))

print(f"Data with parity bits: {''.join(final)}")

# Checker
received_data = input("Enter received data: ")
noOfParity = returnParityBits(len(received_data))
print(f"Number of parity bits: {noOfParity}")
parity = int(input("Enter parity (0 for even, 1 for odd): "))
error = 0

for i in range(noOfParity):
	if generateParityBits(received_data, i, parity):
		error += 2**i

if error:
	print(f"Error at position {error} (from right)")
	received_data = list(received_data)
	received_data[-error] = str(int(received_data[-error]) ^ 1)
	received_data = ''.join(received_data)
	print(f"Corrected transmission sequence: {received_data}")
else:
	print("No error")